drivers/bluetooth/hci/hci_nxp_setup.c

```diff
@@ -0,0 +1,1178 @@
+		return 0;
+	}
+
+	len = sys_get_le16(buffer);
+	len_comp = sys_get_le16(buffer);
+
+	if ((len ^ len_comp) == 0xFFFF) {
+		LOG_DBG("remote asks for %d bytes", len);
+
+		/* Successful. Send back the ack. */
+		if ((fw_upload.hdr_sig == V1_HEADER_DATA_REQ) ||
+		    (fw_upload.hdr_sig == V1_START_INDICATION)) {
+			ack = V1_REQUEST_ACK;
+			fw_upload_write_data(&ack, 1);
+			if (fw_upload.hdr_sig == V1_START_INDICATION) {
+				/* Eliminated longjmp(resync, 1); returning restart status */
+				return (uint16_t)V1_START_INDICATION;
+			}
+		}
+	} else {
+		LOG_ERR("remote asks len %d bytes", len);
+		LOG_ERR("remote asks len_comp %d bytes", len_comp);
+		/* Failure due to mismatch. */
+		ack = 0xbf;
+		fw_upload_write_data(&ack, 1);
+		/* Start all over again. */
+		if (flag) {
+			/* Eliminated longjmp(resync, 1); returning restart status */
+			return (uint16_t)V1_START_INDICATION;
+		}
+		len = 0;
+	}
+	return len;
+}
+
+static uint32_t fw_upload_get_payload_length(uint8_t *cmd)
+{
+	uint32_t len;
+
+	len = sys_get_le32(&cmd[8]);
+
+	return len;
+}
+
+static void fw_upload_get_hdr_start(uint8_t *buffer)
+{
+	int err;
+	bool done = false;
+	uint32_t count = 0;
+
+	while (!done) {
+		err = fw_upload_read_data(&fw_upload.hdr_sig, 1);
+		if (err >= 0) {
+			if (fw_upload.hdr_sig == V1_HEADER_DATA_REQ) {
+				buffer[count++] = fw_upload.hdr_sig;
+				done = true;
+				LOG_DBG("Found header %x", fw_upload.hdr_sig);
+			}
+		}
+	}
+	err = fw_upload_read_data(&buffer[count], 4);
+	if (err > 0) {
+		count += err;
+	}
+}
+
+static int fw_upload_len_valid(uint8_t *buffer, uint16_t *length)
+{
+	uint16_t len;
+	uint16_t len_comp;
+
+	len = sys_get_le16(&buffer[1]);
+	len_comp = sys_get_le16(&buffer[3]);
+
+	if ((len ^ len_comp) == 0xFFFFU) {
+		*length = len;
+		return 0;
+	} else {
+		return -EINVAL;
+	}
+}
+
+static int fw_upload_get_last_5bytes(uint8_t *buffer)
+{
+	int err;
+	uint32_t payload_len;
+	uint16_t len = 0;
+
+	memset(fw_upload.last_5bytes_buffer, 0, sizeof(fw_upload.last_5bytes_buffer));
+
+	fw_upload_get_hdr_start(fw_upload.last_5bytes_buffer);
+	err = fw_upload_len_valid(fw_upload.last_5bytes_buffer, &len);
+	if (err == 0) {
+		LOG_DBG("Valid len %d", len);
+	}
+
+	payload_len = fw_upload_get_payload_length(buffer);
+
+	if ((len == CMD_HDR_LEN) || ((uint32_t)len == payload_len)) {
+		LOG_DBG("Len valid");
+		fw_upload.is_error_case = false;
+		return 0;
+	}
+
+	LOG_DBG("Len invalid");
+	fw_upload.is_error_case = true;
+
+	return -EINVAL;
+}
+
+static void fw_upload_update_result(uint32_t payload_len, uint16_t *sending_len,
+				    bool *first_chunk_sent)
+{
+	if (fw_upload.is_cmd7_req || fw_upload.is_entry_point_req) {
+		*sending_len = CMD_HDR_LEN;
+		*first_chunk_sent = true;
+	} else {
+		*sending_len = payload_len;
+		*first_chunk_sent = false;
+		if (*sending_len == CMD_HDR_LEN) {
+			fw_upload.is_hdr_data = true;
+		}
+	}
+}
+
+static int fw_upload_write_hdr_and_data(uint16_t len_to_send, uint8_t *buffer, bool new_speed)
```

Comment:
A general remark, but I'll take this function as an example;

The number of indentations and possible flows, makes it hard to review IMO. Maybe the process is just that complex, but reading the code doesn't make it easier.

Reply:
Change `fw_upload_write_hdr_and_data` to `fw_upload_write_hdr_and_payload`.

For the function name of `drivers/bluetooth/hci/hci_nxp_setup.c`, it is determined by the firmware download protocol of the controller. The naming is more consistent with the protocol.

Reply:
I didn't mean the name of the function actually, but rather the body...